plugins/janus_sipre.c

```diff
@@ -0,0 +1,2893 @@
+			}
+			if(janus_sipre_allocate_local_ports(session) < 0) {
+				JANUS_LOG(LOG_ERR, "Could not allocate RTP/RTCP ports\n");
+				janus_sdp_free(parsed_sdp);
+				error_code = JANUS_SIPRE_ERROR_IO_ERROR;
+				g_snprintf(error_cause, 512, "Could not allocate RTP/RTCP ports");
+				goto error;
+			}
+			char *sdp = janus_sipre_sdp_manipulate(session, parsed_sdp, FALSE);
+			if(sdp == NULL) {
+				JANUS_LOG(LOG_ERR, "Could not allocate RTP/RTCP ports\n");
+				janus_sdp_free(parsed_sdp);
+				error_code = JANUS_SIPRE_ERROR_IO_ERROR;
+				g_snprintf(error_cause, 512, "Could not allocate RTP/RTCP ports");
+				goto error;
+			}
+			/* Take note of the SDP (may be useful for UPDATEs or re-INVITEs) */
+			janus_sdp_free(session->sdp);
+			session->sdp = parsed_sdp;
+			JANUS_LOG(LOG_VERB, "Prepared SDP for INVITE:\n%s", sdp);
+			/* Prepare the From header */
+			char from_hdr[1024];
+			if(session->account.display_name) {
+				g_snprintf(from_hdr, sizeof(from_hdr), "\"%s\" <%s>", session->account.display_name, session->account.identity);
+			} else {
+				g_snprintf(from_hdr, sizeof(from_hdr), "%s", session->account.identity);
+			}
+			/* Prepare the stack */
+				/* TODO */
+			g_atomic_int_set(&session->hangingup, 0);
+			session->status = janus_sipre_call_status_inviting;
+			/* Create a random call-id */
+			char callid[24];
+			janus_sipre_random_string(24, (char *)&callid);
+			/* Also notify event handlers */
+			if(notify_events && gateway->events_is_enabled()) {
+				json_t *info = json_object();
+				json_object_set_new(info, "event", json_string("calling"));
+				json_object_set_new(info, "callee", json_string(uri_text));
+				json_object_set_new(info, "call-id", json_string(callid));
+				json_object_set_new(info, "sdp", json_string(sdp));
+				gateway->notify_event(&janus_sipre_plugin, session->handle, info);
+			}
+			/* Send INVITE */
+			session->callee = g_strdup(uri_text);
+			session->callid = g_strdup(callid);
+			g_hash_table_insert(callids, session->callid, session);
+			session->media.autoack = do_autoack;
+				/* TODO Use re to send INVITE */
+			g_free(sdp);
+			if(session->transaction)
+				g_free(session->transaction);
+			session->transaction = msg->transaction ? g_strdup(msg->transaction) : NULL;
+			/* Send an ack back */
+			result = json_object();
+			json_object_set_new(result, "event", json_string("calling"));
+		} else if(!strcasecmp(request_text, "accept")) {
+			if(session->status != janus_sipre_call_status_invited) {
+				JANUS_LOG(LOG_ERR, "Wrong state (not invited? status=%s)\n", janus_sipre_call_status_string(session->status));
+				error_code = JANUS_SIPRE_ERROR_WRONG_STATE;
+				g_snprintf(error_cause, 512, "Wrong state (not invited? status=%s)", janus_sipre_call_status_string(session->status));
+				goto error;
+			}
+			if(session->callee == NULL) {
+				JANUS_LOG(LOG_ERR, "Wrong state (no caller?)\n");
+				error_code = JANUS_SIPRE_ERROR_WRONG_STATE;
+				g_snprintf(error_
```

Comment:
re_cancel must be called from the same thread as is running the re_main loop.
you can use struct mqueue to achieve this.